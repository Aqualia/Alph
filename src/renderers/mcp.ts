export type Transport = 'http' | 'sse' | 'stdio';

export interface RenderInput {
  agent: 'cursor' | 'gemini' | 'claude';
  serverId: string;
  transport: Transport;
  url?: string;
  headers?: Record<string, string>;
  command?: string;
  args?: string[];
  cwd?: string;
  env?: Record<string, string>;
  timeout?: number;
}

export interface RenderOutput {
  [containerKey: string]: Record<string, Record<string, unknown>>;
}

function nonEmpty(obj?: Record<string, unknown>): Record<string, unknown> | undefined {
  if (!obj) return undefined;
  const keys = Object.keys(obj);
  return keys.length > 0 ? obj : undefined;
}

export function renderMcpServer(input: RenderInput): RenderOutput {
  const { agent, serverId, transport } = input;

  let server: Record<string, unknown> = {};

  if (agent === 'cursor') {
    if (transport === 'stdio') {
      server = {
        ...(input.command ? { command: input.command } : {}),
        ...(input.args ? { args: input.args } : {}),
        ...(input.env ? { env: input.env } : {})
      };
    } else if (transport === 'sse') {
      server = {
        type: 'sse',
        ...(input.url ? { url: input.url } : {}),
        ...(nonEmpty(input.headers) ? { headers: input.headers } : {})
      };
    } else {
      server = {
        type: 'http',
        ...(input.url ? { url: input.url } : {}),
        ...(nonEmpty(input.headers) ? { headers: input.headers } : {})
      };
    }
  } else if (agent === 'gemini') {
    if (transport === 'stdio') {
      server = {
        transport: 'stdio',
        ...(input.command ? { command: input.command } : {}),
        ...(input.args ? { args: input.args } : {}),
        ...(input.cwd ? { cwd: input.cwd } : {}),
        ...(nonEmpty(input.env) ? { env: input.env } : {}),
        ...(typeof input.timeout === 'number' ? { timeout: input.timeout } : {})
      };
    } else if (transport === 'sse') {
      server = {
        transport: 'sse',
        ...(input.url ? { url: input.url } : {}),
        ...(nonEmpty(input.headers) ? { headers: input.headers } : {}),
        ...(nonEmpty(input.env) ? { env: input.env } : {}),
        ...(typeof input.timeout === 'number' ? { timeout: input.timeout } : {})
      };
    } else {
      server = {
        ...(input.url ? { httpUrl: input.url } : {}),
        ...(nonEmpty(input.headers) ? { headers: input.headers } : {}),
        ...(nonEmpty(input.env) ? { env: input.env } : {}),
        ...(typeof input.timeout === 'number' ? { timeout: input.timeout } : {})
      };
    }
  } else if (agent === 'claude') {
    if (transport === 'stdio') {
      server = {
        ...(input.command ? { command: input.command } : {}),
        ...(input.args ? { args: input.args } : {}),
        ...(nonEmpty(input.env) ? { env: input.env } : {})
      };
    } else if (transport === 'sse') {
      server = {
        type: 'sse',
        ...(input.url ? { url: input.url } : {}),
        ...(nonEmpty(input.headers) ? { headers: input.headers } : {})
      };
    } else {
      server = {
        type: 'http',
        ...(input.url ? { url: input.url } : {}),
        ...(nonEmpty(input.headers) ? { headers: input.headers } : {})
      };
    }
  } else {
    // Fallback: generic shape similar to gemini
    server = transport === 'stdio'
      ? { command: input.command, ...(input.args ? { args: input.args } : {}), ...(nonEmpty(input.env) ? { env: input.env } : {}) }
      : { url: input.url, ...(nonEmpty(input.headers) ? { headers: input.headers } : {}) };
  }

  return {
    mcpServers: {
      [serverId]: server
    }
  };
}

