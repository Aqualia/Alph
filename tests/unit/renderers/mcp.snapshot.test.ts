import { renderMcpServer } from '../../../src/renderers/mcp';

describe('Protocol-aware rendering snapshots', () => {
  const hdr = { Authorization: 'Bearer ${TOKEN}' };
  const env = { FOO: 'bar' };

  test('Cursor STDIO', () => {
    expect(
      renderMcpServer({ agent: 'cursor', serverId: 'github', transport: 'stdio', command: 'npx', args: ['-y','@mcp/github'], env })
    ).toMatchSnapshot();
  });

  test('Cursor SSE', () => {
    expect(
      renderMcpServer({ agent: 'cursor', serverId: 'linear', transport: 'sse', url: 'https://mcp.linear.app/sse', headers: hdr })
    ).toMatchSnapshot();
  });

  test('Cursor HTTP', () => {
    expect(
      renderMcpServer({ agent: 'cursor', serverId: 'notion', transport: 'http', url: 'https://mcp.notion.com/mcp', headers: hdr })
    ).toMatchSnapshot();
  });

  test('Gemini STDIO', () => {
    expect(
      renderMcpServer({ agent: 'gemini', serverId: 'github', transport: 'stdio', command: 'github-mcp', args: [], env, timeout: 5000 })
    ).toMatchSnapshot();
  });

  test('Gemini SSE', () => {
    expect(
      renderMcpServer({ agent: 'gemini', serverId: 'linear', transport: 'sse', url: 'https://mcp.linear.app/sse', headers: hdr, env })
    ).toMatchSnapshot();
  });

  test('Gemini HTTP', () => {
    expect(
      renderMcpServer({ agent: 'gemini', serverId: 'notion', transport: 'http', url: 'https://mcp.notion.com/mcp', headers: hdr })
    ).toMatchSnapshot();
  });

  test('Claude STDIO', () => {
    expect(
      renderMcpServer({ agent: 'claude', serverId: 'local', transport: 'stdio', command: 'claude', args: ['mcp','serve'], env })
    ).toMatchSnapshot();
  });

  test('Claude SSE', () => {
    expect(
      renderMcpServer({ agent: 'claude', serverId: 'linear', transport: 'sse', url: 'https://mcp.linear.app/sse', headers: hdr })
    ).toMatchSnapshot();
  });

  test('Claude HTTP', () => {
    expect(
      renderMcpServer({ agent: 'claude', serverId: 'notion', transport: 'http', url: 'https://mcp.notion.com/mcp', headers: hdr })
    ).toMatchSnapshot();
  });

  // Windsurf
  test('Windsurf STDIO', () => {
    expect(
      renderMcpServer({ agent: 'windsurf', serverId: 'local', transport: 'stdio', command: 'npx', args: ['-y','@mcp/local'], env })
    ).toMatchSnapshot();
  });
  test('Windsurf HTTP', () => {
    expect(
      renderMcpServer({ agent: 'windsurf', serverId: 'remote', transport: 'http', url: 'https://api.example.com/mcp', headers: hdr })
    ).toMatchSnapshot();
  });
});

