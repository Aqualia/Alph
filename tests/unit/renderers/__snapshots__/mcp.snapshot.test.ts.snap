// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Protocol-aware rendering snapshots Claude HTTP 1`] = `
{
  "mcpServers": {
    "notion": {
      "headers": {
        "Authorization": "Bearer \${TOKEN}",
      },
      "type": "http",
      "url": "https://mcp.notion.com/mcp",
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Claude SSE 1`] = `
{
  "mcpServers": {
    "linear": {
      "headers": {
        "Authorization": "Bearer \${TOKEN}",
      },
      "type": "sse",
      "url": "https://mcp.linear.app/sse",
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Claude STDIO 1`] = `
{
  "mcpServers": {
    "local": {
      "args": [
        "mcp",
        "serve",
      ],
      "command": "claude",
      "env": {
        "FOO": "bar",
      },
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Cursor HTTP 1`] = `
{
  "mcpServers": {
    "notion": {
      "headers": {
        "Authorization": "Bearer \${TOKEN}",
      },
      "type": "http",
      "url": "https://mcp.notion.com/mcp",
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Cursor SSE 1`] = `
{
  "mcpServers": {
    "linear": {
      "headers": {
        "Authorization": "Bearer \${TOKEN}",
      },
      "type": "sse",
      "url": "https://mcp.linear.app/sse",
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Cursor STDIO 1`] = `
{
  "mcpServers": {
    "github": {
      "args": [
        "-y",
        "@mcp/github",
      ],
      "command": "npx",
      "env": {
        "FOO": "bar",
      },
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Gemini HTTP 1`] = `
{
  "mcpServers": {
    "notion": {
      "headers": {
        "Authorization": "Bearer \${TOKEN}",
      },
      "httpUrl": "https://mcp.notion.com/mcp",
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Gemini SSE 1`] = `
{
  "mcpServers": {
    "linear": {
      "env": {
        "FOO": "bar",
      },
      "headers": {
        "Authorization": "Bearer \${TOKEN}",
      },
      "transport": "sse",
      "url": "https://mcp.linear.app/sse",
    },
  },
}
`;

exports[`Protocol-aware rendering snapshots Gemini STDIO 1`] = `
{
  "mcpServers": {
    "github": {
      "args": [],
      "command": "github-mcp",
      "env": {
        "FOO": "bar",
      },
      "timeout": 5000,
      "transport": "stdio",
    },
  },
}
`;
